spark/src/test/scala/org/apache/comet/CometCastSuite.scala

```diff
@@ -48,59 +60,118 @@ class CometCastSuite extends CometTestBase with AdaptiveSparkPlanHelper {
 
   ignore("cast string to timestamp") {
-    castTest(Seq("2020-01-01T12:34:56.123456", "T2").toDF("a"), DataTypes.TimestampType)
-    fuzzCastFromString("0123456789/:T \t\r\n", 32, DataTypes.TimestampType)
+    val values = Seq("2020-01-01T12:34:56.123456", "T2") ++ generateStrings(timestampPattern, 8)
+    castTest(values.toDF("a"), DataTypes.DoubleType)
   }
 
-  private def generateFloats = {
+  private def generateFloats(): DataFrame = {
     val r = new Random(0)
-    Range(0, 10000).map(_ => r.nextFloat()).toDF("a")
+    Range(0, dataSize).map(_ => r.nextFloat()).toDF("a")
   }
 
-  private def generateLongs = {
+  private def generateLongs(): DataFrame = {
     val r = new Random(0)
-    Range(0, 10000).map(_ => r.nextLong()).toDF("a")
+    Range(0, dataSize).map(_ => r.nextLong()).toDF("a")
   }
 
-  private def genString(r: Random, chars: String, maxLen: Int): String = {
+  private def generateString(r: Random, chars: String, maxLen: Int): String = {
     val len = r.nextInt(maxLen)
     Range(0, len).map(_ => chars.charAt(r.nextInt(chars.length))).mkString
   }
 
-  private def fuzzCastFromString(chars: String, maxLen: Int, toType: DataType): Unit = {
+  private def generateStrings(chars: String, maxLen: Int): Seq[String] = {
     val r = new Random(0)
-    val inputs = Range(0, 10000).map(_ => genString(r, chars, maxLen))
-    castTest(inputs.toDF("a"), toType)
+    Range(0, dataSize).map(_ => generateString(r, chars, maxLen))
   }
 
   private def castTest(input: DataFrame, toType: DataType): Unit = {
     withTempPath { dir =>
-      val df = roundtripParquet(input, dir)
-        .withColumn("converted", col("a").cast(toType))
-      checkSparkAnswer(df)
+      val data = roundtripParquet(input, dir).coalesce(1)
+      data.createOrReplaceTempView("t")
+
+      withSQLConf((SQLConf.ANSI_ENABLED.key, "false")) {
+        // cast() should return null for invalid inputs when ansi mode is disabled
+        val df = data.withColumn("converted", col("a").cast(toType))
+        checkSparkAnswer(df)
+
+        // try_cast() should always return null for invalid inputs
+        val df2 = spark.sql(s"select try_cast(a as ${toType.sql}) from t")
+        checkSparkAnswer(df2)
+      }
+
+      // with ANSI enabled, we should produce the same exception as Spark
+      withSQLConf(
+        (SQLConf.ANSI_ENABLED.key, "true"),
+        (CometConf.COMET_ANSI_MODE_ENABLED.key, "true")) {
+
+        // cast() should throw exception on invalid inputs when ansi mode is enabled
+        val df = data.withColumn("converted", col("a").cast(toType))
+        val (expected, actual) = checkSparkThrows(df)
+
+        if (CometSparkSessionExtensions.isSpark34Plus) {
+          // We have to workaround https://github.com/apache/datafusion-comet/issues/293 here by
+          // removing the "Execution error: " error message prefix that is added by DataFusion
+          val cometMessage = actual.getMessage
+            .substring("Execution error: ".length)
```

Comment:
Follow on issue to resolve this: https://github.com/apache/datafusion-comet/issues/293